src/coreclr/jit/fgprofile.cpp

```diff
@@ -2190,6 +2250,87 @@ class HandleHistogramProbeInserter
+        , m_instrCount(instrCount)
+    {
+    }
+
+    void operator()(Compiler* compiler, GenTree* node)
+    {
+        if (*m_currentSchemaIndex >= (int)m_schema.size())
+        {
+            return;
+        }
+
+        // Only Buffer.Memmove call is currently expected
+        assert(node->IsCall() && (node->AsCall()->IsSpecialIntrinsic(compiler, NI_System_Buffer_Memmove)));
+
+        const ICorJitInfo::PgoInstrumentationSchema& countEntry = m_schema[*m_currentSchemaIndex];
+        if (countEntry.ILOffset !=
+            static_cast<int32_t>(node->AsCall()->gtHandleHistogramProfileCandidateInfo->ilOffset))
+        {
+            return;
+        }
+
+        bool is32 = countEntry.InstrumentationKind == ICorJitInfo::PgoInstrumentationKind::ValueHistogramIntCount;
+        bool is64 = countEntry.InstrumentationKind == ICorJitInfo::PgoInstrumentationKind::ValueHistogramLongCount;
+        if (!is32 && !is64)
+        {
+            return;
+        }
+
+        assert(*m_currentSchemaIndex + 2 <= (int)m_schema.size());
+        const ICorJitInfo::PgoInstrumentationSchema& tableEntry = m_schema[*m_currentSchemaIndex + 1];
+        assert((tableEntry.InstrumentationKind == ICorJitInfo::PgoInstrumentationKind::ValueHistogram));
+        uint8_t* hist = &m_profileMemory[countEntry.Offset];
+        assert(hist != nullptr);
+
+        *m_currentSchemaIndex += 2;
+
+        GenTree** lenArgRef = &node->AsCall()->gtArgs.GetUserArgByIndex(2)->EarlyNodeRef();
+
+        // We have Memmove(dst, src, len) and we want to insert a call to CORINFO_HELP_VALUEPROFILE for the len:
+        //
+        //  \--*  COMMA     long
+        //     +--*  CALL help void   CORINFO_HELP_VALUEPROFILE
+        //     |  +--*  COMMA     long
+        //     |  |  +--*  STORE_LCL_VAR long  tmp
+        //     |  |  |  \--*  (node to poll)
+        //     |  |  \--*  LCL_VAR   long   tmp
+        //     |  \--*  CNS_INT   long   <hist>
+        //     \--*  LCL_VAR   long   tmp
+        //
+
+        const var_types lenType              = (*lenArgRef)->TypeGet();
+        const unsigned  lenTmpNum            = compiler->lvaGrabTemp(true DEBUGARG("length histogram profile tmp"));
+        compiler->lvaTable[lenTmpNum].lvType = lenType;
+
+        GenTree*     lengthLocal    = compiler->gtNewLclvNode(lenTmpNum, lenType);
+        GenTree*     storeLenToTemp = compiler->gtNewStoreLclVarNode(lenTmpNum, *lenArgRef);
+        GenTreeOp*   lengthNode     = compiler->gtNewOperNode(GT_COMMA, lenType, storeLenToTemp, lengthLocal);
+        GenTree*     histNode       = compiler->gtNewIconNode(reinterpret_cast<ssize_t>(hist), TYP_I_IMPL);
+        unsigned     helper         = is32 ? CORINFO_HELP_VALUEPROFILE32 : CORINFO_HELP_VALUEPROFILE64;
+        GenTreeCall* helperCallNode = compiler->gtNewHelperCallNode(helper, TYP_VOID, lengthNode, histNode);
+
+        *lenArgRef = compiler->gtNewOperNode(GT_COMMA, lenType, helperCallNode, compiler->gtCloneExpr(lengthLocal));
```

Comment:
This doesn't seem like it handles small types correctly.

```suggestion
        const unsigned  lenTmpNum            = compiler->lvaGrabTemp(true DEBUGARG("length histogram profile tmp"));
        GenTree*     storeLenToTemp = compiler->gtNewTempStore(lenTmpNum, *lenArgRef);
        GenTree*     lengthLocal    = compiler->gtNewLclvNode(lenTmpNum, genActualType(*lenArgRef));
        GenTreeOp*   lengthNode     = compiler->gtNewOperNode(GT_COMMA, lengthLocal->TypeGet(), storeLenToTemp, lengthLocal);
        GenTree*     histNode       = compiler->gtNewIconNode(reinterpret_cast<ssize_t>(hist), TYP_I_IMPL);
        unsigned     helper         = is32 ? CORINFO_HELP_VALUEPROFILE32 : CORINFO_HELP_VALUEPROFILE64;
        GenTreeCall* helperCallNode = compiler->gtNewHelperCallNode(helper, TYP_VOID, lengthNode, histNode);

        *lenArgRef = compiler->gtNewOperNode(GT_COMMA, lengthLocal->TypeGet(), helperCallNode, compiler->gtCloneExpr(lengthLocal));
```

Reply:
Thanks! Currently we don't poll small-typed values but I guess it's good to make it more generic for future uses

Reply:
Won't we still poll e.g. a `ref ushort x` loaded and passed as the length? This would create some unusual `TYP_USHORT` nodes in that case.
It should only happen on x86/arm32 since `Buffer.Memmove` takes a `nuint`.

Reply:
ah I see